app/grandchallenge/retina_api/views.py

```diff
@@ -896,3 +905,137 @@ class RetinaImagePathologyAnnotationViewSet(viewsets.ModelViewSet):
     filter_backends = (filters.DjangoObjectPermissionsFilter,)
     pagination_class = None
     queryset = RetinaImagePathologyAnnotation.objects.all()
+
+
+class ArchiveAPIView(APIView):
+    permission_classes = (RetinaAPIPermission,)
+    authentication_classes = (authentication.TokenAuthentication,)
+    renderer_classes = (JSONRenderer,)
+    pagination_class = None
+
+    @staticmethod
+    def create_response_object(archives, patients):
+        def generate_archives(archive_list, patients):
+            for archive in archive_list:
+                archive_only_images = []
+                for image in archive.images.all():
+                    if image.study is None:
+                        archive_only_images.append(image)
+
+                images = sorted(
+                    generate_images(archive_only_images), key=lambda x: x["id"]
+                )
+                subfolders = sorted(
+                    generate_patients(archive, patients), key=lambda x: x["id"]
+                )
+
+                yield {
+                    "id": archive.id,
+                    "name": archive.name,
+                    "subfolders": subfolders,
+                    "images": images,
+                }
+
+        def generate_patients(archive, patients):
+            patient_list = patients.filter(
+                study__image__archive=archive
+            ).distinct()
+            for patient in patient_list:
+                yield {
+                    "id": patient.id,
+                    "name": patient.name,
+                    "subfolders": sorted(
+                        generate_studies(patient.study_set),
+                        key=lambda x: x["id"],
+                    ),
+                    "images": [],
+                }
+
+        def generate_studies(study_list):
+            for study in study_list.all():
+                yield {
+                    "id": study.id,
+                    "name": study.name,
+                    "images": sorted(
+                        generate_images(study.image_set.all()),
+                        key=lambda x: x["id"],
+                    ),
+                    "subfolders": [],
+                }
+
+        def generate_images(image_list):
+            for image in image_list:
+                yield {
+                    "id": image.id,
+                    "name": image.name,
+                    "eye": image.eye_choice,
+                    "modality": image.modality.modality,
+                    "patient": image.study.patient.name,
+                    "study": image.study.name,
+                    "archives": image.archive_set.values(),
+                }
+
+        return sorted(
+            generate_archives(archives, patients), key=lambda x: x["id"]
+        )
+
+    def get(self, request):
+        archives = Archive.objects.all().prefetch_related(
```

Comment:
I believe it can't be used for ManyToMany relations, which the Archive-Image relation is. Also not for reverse ForeignKey relations which the second lookup is (patients-study-image).

I did just notice I don't use the obs_image and oct_image so I removed that.